full-stack-ruby-on-rails/project-binary-search-trees/lib/tree.rb

```diff
@@ -0,0 +1,195 @@
+# frozen_string_literal: true
+
+require_relative './node'
+
+# A class which constructs a binary search tree from an array
+class Tree
+  attr_reader :root
+
+  def initialize(array = [])
+    @root = build_tree(array)
+  end
+
+  def build_tree(array)
+    array = sanitize_array(array)
+    build_branches(array, start_index: 0, end_index: array.size-1)
+  end
+
+  def build_branches(array, start_index:, end_index:)
+    return nil if start_index > end_index
+
+    mid_point = (start_index + end_index) / 2
+
+    root = Node.new(array[mid_point])
+    root.left = build_branches(array,
+                               start_index: start_index,
+                               end_index: mid_point - 1)
+
+    root.right = build_branches(array,
+                                start_index: mid_point + 1,
+                                end_index: end_index)
+
+    root
+  end
+
+  def insert(value, node = @root)
+    return Node.new(value) if node.nil?
```

Comment:
Minor: I typically see it written this way:

```suggestion
    return Node.new(value) unless node
```

Just an observation. Dunno if Rubocop cares about this stuff.